casper/src/main/rholang/MakeRev.rho

```diff
@@ -0,0 +1,177 @@
+  transferIfEnough, checkedAddSubtract
+in {
+  contract exampleUsage(_) = {
+    new revCh, contract1, contract2, c3, c4, c5, trace(`rho:io:stderr`) in {
+      MakeRev!({
+        bundle0{*contract1}: 100,
+        bundle0{*c3}: 9223372036854775807
+      }, *revCh) |
+      for (rev <- revCh) {
+
+        // Simple send from one contract "account" to another.
+        new ch, ch2, ch3, ch4, ch6 in {
+          rev!("getBalance", bundle0{*contract1}, *ch) | for(@bc1 <- ch) {
+            trace!({"contract1 balance before": bc1}) |
+            rev!("sendNameToName", bundle0{*contract1}, bundle0{*contract2}, 75, *ch) |
+            for(@sendResult <- ch) {
+              trace!({"sendResult": sendResult}) |
+
+              // Overdraft
+              rev!("sendNameToName", bundle0{*contract1}, bundle0{*contract2}, 100, *ch) |
+              for(@sendResult <- ch) {
+                trace!({"2nd sendResult": sendResult})
+              }
+            } |
+
+            // Overflow
+            rev!("sendNameToName", bundle0{*contract1}, bundle0{*c3}, 10, *ch2) |
+            for(@sendResult <- ch2) {
+              trace!({"big sendResult": sendResult})
+            } |
+
+            // Junk destination
+            rev!("sendNameToKey", bundle0{*contract1}, Nil, 10, *ch3) |
+            for(@sendResult <- ch3) {
+              trace!({"junk dest sendResult": sendResult})
+            } |
+            rev!("sendNameToKey", bundle0{*contract1}, "0xdeadbeef", 10, *ch4) |
+            for(@sendResult <- ch4) {
+              trace!({"short dest sendResult": sendResult})
+            } |
+
+            // send to key
+            rev!("sendNameToKey", bundle0{*contract1},
+              "0x02c50804457ff616801953fa72672564f40a58d9", 2, *ch6) |
+            for(@sendResult <- ch6) {
+              trace!({"send to key result": sendResult})
+            }
+          }
+        }
+      }
+    }
+  } |
+  exampleUsage!(Nil) |
+
+  // This morning (Jan 24) I said this contract is still a mint
+  // in that whoever holds it can print money.
+  // I have reconsidered; it can be published in the
+  // registry and anyone can call its methods, which
+  // conserve money. So I'm calling it MakeRev.
+  contract MakeRev(@initialBalances, return) = {
+    new self, balancesCh, transfer in {
+      return!(bundle+{*self}) |
+
+      // ISSUE: should be a concurrent map
+      balancesCh!(initialBalances) |
+
+      contract self(@"getBalance", @holder, return) = {
+        for (@balances <- balancesCh) {
+          balancesCh!(balances) | // (should be peek)
+          return!(balances.get(holder))
+        }
+      } |
+
+      contract self(@"sendKeyToKey",
+        @{from /\ String}, // sender's public key
+                           // ISSUE: public key or hash of key ala ethAddr?
+                           // ISSUE cryptoSuite: secp256k1 or ed25591? secp256r1? parameterized?
+        @nonce,            // Distinctness is the responsibility of the signer;
```

Comment:
It doesn't matter if the nonce is serial or not. The important thing is that the receiver has to be aware of the next valid nonce, otherwise the attacker will simply replay the call with a previously used pair of nonce&signature

Reply:
"next" valid nonce? the nonces need not come in any order; any fresh name from `new x in { ...` will do.

Reply:
oh.. the *receiver* has to reject previously used nonces... yes, of course.

back to the drawing board...

Reply:
Now my tiny brain sees how the proxy helps...

Reply:
1a8e4a4 includes `LockBox` that uses a proxy.